### PR TITLE
Allow active job classes

### LINKED ISSUE #42: Scheduling with ActiveJob
I am trying to use que-scheduler with Rails and ActiveJob. When I specify a subclass of ActiveJob::Base in my schedule, [this subclass check](https://github.com/hlascelles/que-scheduler/blob/master/lib/que/scheduler/defined_job.rb#L19) fails. Any hint what to do?

### PR DESCRIPTION
**Pull request by papodaca**
A few minor tweaks that are required to get ActiveJob based scheduled tasks working.

Closes #42

### CODE REVIEW

**Comment by jglauche** [comment]
I've been running your fork, it seems to work but spits out a warning in the log file

{"lib":"que","hostname":"meow","pid":24423,"thread":46983057656680,"event":"que-scheduler","message":"que-scheduler called enqueue on AlertsJob but did not receive a Que::Job"}

my AlertsJob is a kind of ApplicationJob

**Comment by papodaca** [comment]
Looks like there is some[ validation logic](https://github.com/hlascelles/que-scheduler/blob/master/lib/que/scheduler/scheduler_job.rb#L56) that expects only a Que::Job to be created.

**Comment by hlascelles** [comment]
Aha... I can give more background on the existence of that.

Internally (work, closed source), we have written `Que::Unique`, which is a plugin / monkey patch to coalesce multiple identical enqueues  in one transaction, to avoid overwork.

eg (pseudo code flow):

```
start_db_transaction
update_customer # => causes "ReindexCustomerStatusQueJob" to be enqueued
... some other logic ...
update_customer_order # => also causes "ReindexCustomerStatusQueJob" to be enqueued
commit_db_transaction
```

Normally, both `ReindexCustomerStatusQueJob` jobs would run afterwards (unnecessarily). 

So, with `Que::Unique` we have patched `.enqueue` to collect multiple identical calls in one transaction and squash them down, so only one of each type is enqueued at the commit. It works very well.... *except* when que-scheduler runs, and tries to enqueue multiples of the same job deliberately (which it might do, depending on your `que-scheduler.yaml` config).

Long story short, that class `is_a?` check is mostly just asserting that the result of an `.enqueue` is not `nil` (something has gone wrong), or a mock job intercept class. 

On your branch, changing it to be `enqueued_job.is_a?(Que::Job) || enqueued_job.is_a?(ApplicationJob)` will probably be the right thing to do (I have never used ActiveJob).

**Comment by papodaca** [comment]
@hlascelles Oops didn't see your message. I've updated this PR to include some more changes to get all the way to audits getting saved.

**Comment by papodaca** [comment]
Its not clear to me if they way I am making he ActiveJob Vs vanilla Que is correct way. This seems to be working correctly so far however.

**Comment by hlascelles** [comment]
Thank you for this. I will get up to speed on ActiveJob so I can add the relevant tests. Might not be for a few weeks, but I will get onto it.

**Comment by Rinkana** [comment]
@hlascelles do you happen to have any update on this? I'd love to try this out on Rails 6 + AJ. If i can help with this please let me know.

@papodaca could you rebase this PR as it's a bit outdated (especially que beta4 support)

**Comment by papodaca** [comment]
Rebase has been performed, luckily there were no conflicts.

**Comment by hlascelles** [comment]
Thank you for looking at implementing this.

I have refactored it on another branch to add tests: https://github.com/hlascelles/que-scheduler/compare/pr-138?expand=1

When catering for `priority` and `queue` being configured in the `que-scheduler.yaml` it does get a little more complex as you can see from the `enqueue` code.

But even over and above that, I cannot get queue *names* to work at all, ie enqueueing a job onto a particular queue. In fact, following the code through, I believe the ActiveJob Que adapter does not support queue names at all. [This line](https://github.com/rails/rails/blob/c70112e74f4d2ef517f4036fe6e2888cc30fc952/activejob/lib/active_job/enqueuing.rb#L51) appears to set `queue_name`, but [this line](https://github.com/rails/rails/blob/c70112e74f4d2ef517f4036fe6e2888cc30fc952/activejob/lib/active_job/queue_adapters/que_adapter.rb#L21) does not use it. I have written [a test](https://github.com/hlascelles/que-scheduler/compare/pr-138?expand=1#diff-0cc32fba3934a89a89deab5c231e9bbfR16) that tries to prove that.

I would be grateful if you could see if that branch still works for your use case. When we have all the issues resolved we can merge it into your PR.

To run all the tests:

```
# In one window:
docker run -p 5430:5432 postgres:9.5.0
# In another:
DB_PORT=5430 DB_PASSWORD=postgres bundle exec appraisal rspec
```

Could you also check my analysis and show if ActiveJob does indeed not support queue names for Que?

**Comment by Rinkana** [comment]
I was browsing trough the history of of activejob and i found [this](https://github.com/rails/rails/commit/fdfca02c304d809aac53ebf8094b237d0ba6415e) commit that removes the named queues option with [this](https://github.com/rails/rails/pull/19498) PR for more information.
TLDR: 

> Named queues were added to Que specifically for the (rare) use case wherein multiple codebases need to use the same database, 
... 
They weren't intended for use as a priority system

**Comment by hlascelles** [comment]
Cheers @Rinkana, I will add documentation saying named queues are not supported for ActiveSupport.

I've done some preliminary refactoring on master to get this PR merged in. I've spent most of my time on the spec side to make sure all the tests provide full coverage. I will produce a branch shortly!

**Comment by hlascelles** [comment]
Ongoing... It's always the minor tweaks that come with significant test refactoring. Here is [the branch](https://github.com/hlascelles/que-scheduler/compare/pr-138?expand=1) so far. 

The overwhelming bulk of the changes is to run the whole test suite twice, once with `Que::Job` classes and once with `ActiveJob`.

I believe the runtime code works (but will need cleanup). It's just two tests that are failing.

I will do some pre-prs to get the framework in place then we can rebase this PR to get it merged in.

**Comment by hlascelles** [comment]
Hmmm... Funnily enough named queues are about to be readded https://github.com/rails/rails/pull/38635

**Comment by hlascelles** [comment]
OK, I have spent a bit of time on this, and it required a full rewrite of some of the internal logic. Thank you for this PR to guide me, I have opened a new one with attribution to you @papodaca 

https://github.com/hlascelles/que-scheduler/pull/164

I'll cut a gem for it when tests pass.

**Comment by hlascelles** [comment]
Released: https://rubygems.org/gems/que-scheduler/versions/3.3.0
